data-serving/scripts/convert-data/geocode_util.py

```diff
@@ -0,0 +1,200 @@
+        geocoder, lambda location: location.admin2.lower() ==
+        higher_res_location_token and location.admin1.lower() ==
+        lower_res_location_token and location.geo_resolution == 'admin2')
+
+    if city_and_province:
+        return city_and_province
+
+    return None
+
+
+def lookup_single_part_location(geocoder: Any, value: str) -> Any:
+    '''
+    Attempts to match the location token against known locations, beginning from
+    lowest resolution and moving to higher resolutions if no match is found. No
+    Paris, TX matches for "Paris" over here.
+
+    Returns:
+      None: When no geocoding match is found.
+      [Dict[str, Any]]: Location data of the form:
+        {
+          lat: float
+          lng: float
+          geo_resolution: str
+          country_new: str
+          admin_id: int
+          location: str
+          admin3: str
+          admin2: str
+          admin1: str
+        }
+    '''
+    country = lookup(geocoder, lambda location: location.country_new.lower()
+                     == value and location.geo_resolution == 'admin0')
+    if country:
+        return country
+
+    province = lookup(geocoder, lambda location: location.admin1.lower()
+                      == value and location.geo_resolution == 'admin1')
+    if province:
+        return province
+
+    city = lookup(geocoder, lambda location: value ==
+                  location.admin2.lower() and location.geo_resolution == 'admin2')
+    if city:
+        return city
+
+    admin3 = lookup(geocoder, lambda location: value ==
+                    location.admin3.lower() and location.geo_resolution == 'admin3')
+    if admin3:
+        return admin3
+
+    location = lookup(geocoder, lambda location: value == location.location.lower(
+    ) and location.geo_resolution == 'point')
+    if location:
+        return location
+
+    return None
+
+
+def lookup(geocoder: Any, predicate: Callable[[str], bool]) -> Any:
+    '''
+    Finds a single geocode match given a predicate. The predicate could, for
+    example, require that the result match a given string and a given
+    resolution. If more than one match is found for the predicate, an error is
+    thrown.
+
+    Raises:
+      ValueError: When more than one match is found for the predicate, since we
+        are unable to determine which is correct.
+    Returns:
+      None: When no geocoding match is found.
+      [Dict[str, Any]]: Location data of the form:
+        {
+          lat: float
+          lng: float
+          geo_resolution: str
+          country_new: str
+          admin_id: int
+          location: str
+          admin3: str
+          admin2: str
+          admin1: str
+        }
+    '''
+    matches = [
+        geocode for id, geocode in geocoder.geocodes.items()
+        if predicate(geocode)]
+
+    if len(matches) > 1:
+        raise ValueError(f'Too many possible geocode matches: {matches}')
+
+    return matches[0] if len(matches) == 1 else None
```

Comment:
Another feel-free-to-ignore [Pythonism](https://docs.python.org/3/library/functions.html#next): `next(iter(matches), None)`

Reply:
thanks!